File: bindings/wasm/examples/basic.mjs
/**
 * Example: Basic usage of the ROX WASM bindings with Node.js.
 * 
 * Build first:
 *   cd bindings/wasm
 *   wasm-pack build --target nodejs
 * 
 * Then run:
 *   node examples/basic.mjs
 */

import { readFileSync, writeFileSync } from 'fs';
import { decode, encode, convert, version } from '../pkg/rox_wasm.js';

console.log(`ROX WASM v${version()}`);
console.log("");

// Decode a Quaver chart
console.log("=== Decoding Quaver chart ===");
const quaData = readFileSync("../../assets/quaver/4K.qua");
const chart = decode(quaData, "qua");

console.log(`Title:      ${chart.title}`);
console.log(`Artist:     ${chart.artist}`);
console.log(`Creator:    ${chart.creator}`);
console.log(`Difficulty: ${chart.difficulty}`);
console.log(`Keys:       ${chart.key_count}K`);
console.log(`Notes:      ${chart.note_count}`);
console.log(`Duration:   ${chart.duration.toFixed(1)}s`);
console.log(`Hash:       ${chart.hash}`);
console.log(`Coop:       ${chart.is_coop}`);
console.log("");

// Encode to osu! format
console.log("=== Converting to osu! ===");
const osuBytes = encode(chart, "osu");
writeFileSync("output_from_node.osu", Buffer.from(osuBytes));
console.log("Saved: output_from_node.osu");
console.log("");

// Decode FNF chart
console.log("=== Decoding FNF chart ===");
const fnfData = readFileSync("../../assets/fnf/test-song.json");
const fnfChart = decode(fnfData, "json");
console.log(`Title: ${fnfChart.title}, ${fnfChart.key_count}K, ${fnfChart.note_count} notes`);
console.log("");

// Direct conversion
console.log("=== Direct conversion ===");
const qua7kData = readFileSync("../../assets/quaver/7K.qua");
const smBytes = convert(qua7kData, "qua", "sm");
writeFileSync("output_7k.sm", Buffer.from(smBytes));
console.log("Converted 7K.qua -> output_7k.sm");

File: bindings/python/examples/basic.py
#!/usr/bin/env python3
"""
Example: Basic usage of the ROX Python bindings.

Run from the bindings/python directory:
    .venv/Scripts/python examples/basic.py
"""

import rox

def main():
    # Decode a Quaver chart
    print("=== Decoding Quaver chart ===")
    chart = rox.decode("../../assets/quaver/4K.qua")
    
    print(f"Title:      {chart.title}")
    print(f"Artist:     {chart.artist}")
    print(f"Creator:    {chart.creator}")
    print(f"Difficulty: {chart.difficulty}")
    print(f"Keys:       {chart.key_count}K")
    print(f"Notes:      {chart.note_count}")
    print(f"Duration:   {chart.duration:.1f}s")
    print(f"Hash:       {chart.hash}")
    print(f"Coop:       {chart.is_coop}")
    print()
    
    # Convert to osu! format
    print("=== Converting to osu! ===")
    rox.encode(chart, "output_from_python.osu")
    print("Saved: output_from_python.osu")
    print()
    
    # Decode FNF chart
    print("=== Decoding FNF chart ===")
    fnf_chart = rox.decode("../../assets/fnf/test-song.json")
    print(f"Title: {fnf_chart.title}, {fnf_chart.key_count}K, {fnf_chart.note_count} notes")
    print()
    
    # Direct conversion
    print("=== Direct conversion ===")
    rox.convert("../../assets/quaver/7K.qua", "output_7k.osu")
    print("Converted 7K.qua -> output_7k.osu")

if __name__ == "__main__":
    main()
